### ISSUE TITLE
Clarifying Valid Types

### ISSUE THREAD

**Issue by saulshanabrook** [egraphs-good/egglog#298]
I am looking to see if I can tighten the static analysis type in the Python bindings and wanted to get some feedback on my current understanding of what is allowed:

* `union`: Must be user-defined sorts
* `set`: Must be primitive sorts
* `delete`: Can be either
* `rewrite`: Must be user-defined sorts
* `=`: Can be either
* `!=`: Can be either, but only sound to use it on primitive sorts, since user-defined sorts could later become `=` even if they aren't at the moment. I would make this only work on primitive sorts

Does that seem right?

**Comment by oflatt** [comment]
That doesn't sound right: I use `set` with non-primitive sorts sometimes.
With `set`, it's important that the function isn't a normal datatype though (has a merge function or default).
I think we should have a different keyword than function for normal datatypes from the `datatype` form.

That seems right besides that.

**Comment by saulshanabrook** [comment]
Thanks @oflatt for the clarification.

I realized I am a little fuzzy on how adding a default or merge function changes the non-primitive sorts. 

My current understanding of `set` is that it changes one row in a function `Table` with a new output `Value` for some set of `Input` values. With non-primitive sorts (`EqSort`s) their value is an index into the `parents` vec of the union-find data structure.

**So I am unclear what the `set` would do in the case where it operates on `EqSort`s and how it would differ from `union`.**

If I try out this example:

```
(datatype Math
  (Num i64)
  (Var String)
  (comb Math Math))

(function special () Math :cost 10 :merge (comb old new) :default (Num 0) )
(special)
(set (special) (Num 10))
(union (special) (Num 5))
```
<img width="467" alt="Screenshot 2023-11-25 at 3 17 44 PM" src="https://github.com/egraphs-good/egglog/assets/1186124/4234469d-5ea7-449c-861b-191894a69d89">

**It looks like the difference is that `set` will end using the `merge` function whereas `union` won't. Is that how you understand the semantics as well?**

Taking a look at the code, this example is desugared as:

```
(sort Math)
(function Num (i64) Math)
(function Var (String) Math)
(function comb (Math Math) Math)
(function special () Math :cost 10 :merge (comb old new) :default (Num 0))
(let v0___ (special))
(let v2___ 10)
(let v1___ (Num v2___))
(set (special) v1___)
(let v3___ (special))
(let v5___ 5)
(let v4___ (Num v5___))
(union v3___ v4___)
```

If we add a `println!("instructions: {:?}", checker.instructions)` into the [`compiler_actions` function](https://github.com/oflatt/egg-smol/blob/f1fad223f5d56a2a5caa89e9f755168d4357bd80/src/typecheck.rs#L725-L726), we see that we end up with `Set` and `Union` instructions, as we might expect:

```
instructions: [Global("v1___"), Set("special")]
instructions: [Global("v3___"), Global("v4___"), Union(2)]
```

The [`Union` instruction will end up calling `unionfind.union`](https://github.com/oflatt/egg-smol/blob/f1fad223f5d56a2a5caa89e9f755168d4357bd80/src/typecheck.rs#L904-L914) whereas [the `Set` instruction will not use the unionfind directly if a custom merge function is set. If it is, it will eval that expression and use that value as the result](https://github.com/oflatt/egg-smol/blob/f1fad223f5d56a2a5caa89e9f755168d4357bd80/src/typecheck.rs#L782-L789).

**Comment by saulshanabrook** [comment]
> Yes, union only works on eqsorts. On the other hand, set uses merge functions.

So this means that if you define a `rewrite` on a function with a custom merge function it won't be triggered if the return type is an eqsort? My assumption would be the reverse.

It makes me wonder if it's necessary to keep separate `set` and `union` commands and if they could just be unified that into one command? Is it ever necessary to use the `union` behavior on eqsat types with merge functions to ignore the merge function?

> I thought we disallowed using set on things that don't have a user-specified merge function to clarify this distinction. Is that not true? If not, we should fix that.

I think it was disallowed, but there was a leftover code path for when it was still there  I was looking at. I just submitted a PR to remove that unused code https://github.com/egraphs-good/egglog/pull/307

**Comment by oflatt** [comment]
Hmm, if you use `rewrite` on a function with a custom merge function, shouldn't you get a type error? If not we need to fix that.

**Comment by saulshanabrook** [comment]
No there is no type error for [a rewrite with a custom merge function](https://egraphs-good.github.io/egglog/?program=XQAAgADGAAAAAAAAAAAUGQgnfMUD9dO10C6gYDgH7EMCDsP59lqRF2h73A5qXxBcqLSuVAZFJFJ6Jma5Gd39gwjEeMRd0n1zwQVR347S9aItyTeMwIBmOMprCBmwpLRYGaEMQsZfMC5uIHRtSc76I2CLU6-Rtd9MhhH8FSIGjKe5z9kovBe6__85CYAA):

```
(datatype Math
  (Num i64)
  (Add Math Math))

(function custom (Math Math) Math :merge (Add old new) :default (Num 0))

(rewrite (custom (Num 1) (Num 2)) (Num 3))

(custom (Num 1) (Num 2))

(run 1)
```

**Comment by oflatt** [comment]
Okay, that's bad. 
This thread is convincing me that we should make this distinction more clear. In egglog, there should be two types of tables:
`datatype` tables and `function` tables.
`datatype` tables can be `union`ed but `function` tables must be `set`.

**Comment by oflatt** [comment]
It also doesn't really make sense to have a `default` with no `merge` function in my mind. It seems like you could just `union` a datatype with your default.

**Comment by oflatt** [comment]
I'll try to come up with an RFC to solve this conundrum

**Comment by saulshanabrook** [comment]
> I'll try to come up with an RFC to solve this conundrum

Thanks! Yeah it's definitely a bit idiosyncratic at the moment. If you want someone to bounce ideas off of happy to jump on a call too.

**Comment by saulshanabrook** [comment]
> * `!=`: Can be either, but only sound to use it on primitive sorts, since user-defined sorts could later become = even if they aren't at the moment. I would make this only work on primitive sorts

I noticed that the lambda example [uses `!=` in some rules on eq sorts](https://github.com/egraphs-good/egglog/blob/41562802315eccc9136c9d2ae6967d2df4071420/tests/lambda.egg#L74), so it seems like I should allow Python users to also call `!=` on their user-defined sorts? Does that seem right?

**Comment by oflatt** [comment]
Created #309, let me know what you think!
Using `!=` on user-defined sorts works- but you have to be very careful! If the inequality changes later, unsoundness might occur, or semi-naiive evaluation may skip evaluating your query (unsoundly).

**Comment by saulshanabrook** [comment]
Thanks, @oflatt, for all the clarifications on what's valid currently! Compared to the initial post I have realized that:

* `!=` should be allowed on all types, not just eqsorts, since it is [used in the lambda example](https://github.com/egraphs-good/egglog/blob/41562802315eccc9136c9d2ae6967d2df4071420/tests/lambda.egg#L74)
* `set` can be used on either primitive or eqsorts. It isn't allowed on eqsorts that have the default merge function, but is allowed on ones with custom merge functions or defaults.

I'll close this since the questions this issue raised have been resolved. They have brought to light some potential points of confusion around the existing implementation, but I will leave that to another issue to discuss if necessary.
